src/file/mod.rs

```diff
@@ -1,35 +1,46 @@
+pub mod extension;
 mod format;
 pub mod source;
 
 use std::collections::HashMap;
+use std::fmt::Debug;
 use std::path::{Path, PathBuf};
 
 use crate::error::*;
 use crate::source::Source;
 use crate::value::Value;
+use crate::Format;
 
 pub use self::format::FileFormat;
 use self::source::FileSource;
 
+pub use self::extension::FileExtensions;
 pub use self::source::file::FileSourceFile;
 pub use self::source::string::FileSourceString;
 
+/// A configuration source backed up by a file.
+///
+/// It supports optional automatic file format discovery.
 #[derive(Clone, Debug)]
-pub struct File<T>
+pub struct File<T, F>
 where
-    T: FileSource,
+    F: Format + FileExtensions,
+    T: FileSource<F>,
 {
     source: T,
 
     /// Format of file (which dictates what driver to use).
-    format: Option<FileFormat>,
+    format: Option<F>,
 
     /// A required File will error if it cannot be found
     required: bool,
 }
 
-impl File<source::string::FileSourceString> {
-    pub fn from_str(s: &str, format: FileFormat) -> Self {
+impl<F> File<source::string::FileSourceString, F>
+where
+    F: Format + FileExtensions + 'static,
```

Comment:
I wonder how the `'static` bound behaves in production code then... :thinking:

Reply:
It only disallows `F` implementations to have borrowed fileds, it can be easily worked around with a simple `clone` or `Rc`. Unless somebody adds sources to config in a hot loop in a critical path, this should be fine.